### PR TITLE
Add seq parallelism for attention and MoE MLP

### PR DESCRIPTION
**Pull request by suexu1025**
# Description

1. Add seq_parallelism + exp_parallelism for attention + followed MLP module
with sp+ep, moe customer 2k seq inference improved by 20%
2. Fix prefill_KV_cache sharding mismatch during seq_parallelism
3. decode improved by 10%
4. Enable inference auto layout in mistral model

FIXES: b/374773995

# Tests

tested on v6e/v5p:
SEQ=2048

python MaxText/inference_microbenchmark.py MaxText/configs/inference.yml max_prefill_predict_length=$SEQ max_target_length=6144 model_name=mixtral-8x7b ici_fsdp_parallelism=1 ici_autoregressive_parallelism=1 ici_expert_parallelism=1 ici_context_parallelism=4 ici_tensor_parallelism=1 scan_layers=false per_device_batch_size=1 attention=dot_product megablox=False quantization=int8 checkpoint_is_quantized=True quantize_kvcache=True capacity_factor=1 tokenizer_path=assets/tokenizer.mistral-v3 compute_axis_order=0,2,1,3 ar_cache_axis_order=0,2,1,3 enable_jax_profiler=True inference_microbenchmark_prefill_lengths="$SEQ" base_output_directory=$OUT_DIR run_name=$RUN_NAME profiler=xplane model_call_mode=inference inference_microbenchmark_stages=prefill

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed.

### CODE REVIEW

File: MaxText/layers/attentions.py [hunk 1]
```diff
@@ -1424,7 +1457,7 @@ def out_projection(self, output_dim: int, out: Array) -> Array:
         features=output_dim,
         axis=(-2, -1),
         kernel_init=self.kernel_init,
-        kernel_axes=("heads", "kv", "embed"),
+        kernel_axes=(None, None, None), # trade speed with memory
```
Comment (mailvijayasingh) [inline MaxText/layers/attentions.py:1460]: does this mean we might OOM in some higher batch sizes?

File: MaxText/layers/mistral.py [hunk 2]
```diff
@@ -103,6 +103,9 @@ def __call__(
         float32_logits=cfg.float32_logits,
         quant=self.quant,
         kv_quant=quantizations.configure_kv_quant(cfg),
+        prefill_cache_axis_order=tuple([int(i) for i in cfg.prefill_cache_axis_order.split(",")]),
+        ar_cache_axis_order=tuple([int(i) for i in cfg.ar_cache_axis_order.split(",")]),
```
Comment (mailvijayasingh) [inline MaxText/layers/mistral.py:107]: thanks for fixing this :)

File: MaxText/pyconfig.py [hunk 3]
```diff
@@ -212,6 +212,7 @@ def validate_model_name(s: str) -> bool:
       "llama3.1-70b",
       "llama3.1-405b",
       "llama3.3-70b",
+      "subsup",
```
Comment (mailvijayasingh) [inline MaxText/pyconfig.py:215]: remove

File: MaxText/layers/linears.py [hunk 4]
```diff
@@ -635,49 +654,74 @@ def dense_matmul(self, inputs, gate_logits, w0_kernel, w1_kernel, wo_kernel):
     weights = self.reshape_and_update_weights(top_k_weights, top_k_indices)
     matmul_precision = lax.Precision(self.config.matmul_precision)
 
+    cp = self.config.ici_context_parallelism
+    batch_size = inputs.shape[0]
+    seq_len = inputs.shape[1]
+    if seq_len % cp != 0:
```
Comment (mailvijayasingh) [inline MaxText/layers/linears.py:660]: nit: maybe abstract this part to get_cp and get_sub_seq_length?

File: MaxText/configs/base.yml [hunk 5]
```diff
@@ -257,58 +257,63 @@ jax_cache_dir: "~/jax_cache"
...
 logical_axis_rules: [
                       ['activation_batch', ['data', 'fsdp', 'fsdp_transpose', 'expert']],
                       ['activation_batch_no_exp', ['data', 'fsdp', 'fsdp_transpose']],
                       ['activation_embed_and_logits_batch', ['data', 'stage', 'fsdp', 'fsdp_transpose', 'expert']],
                       ['activation_heads', ['tensor', 'tensor_transpose', 'sequence','tensor_sequence']],
                       ['activation_kv_heads', ['tensor', 'tensor_transpose', 'sequence','tensor_sequence']],
-                      ['activation_length', ['sequence']],
+                      ['activation_length', ['sequence', 'context']],
+                      ['activation_length', ['context']],
+                      ['activation_length_q', ['context']],
```
Comment (mailvijayasingh) [inline MaxText/configs/base.yml:269]: as talked offline, shall we pull out a config, moe_config_inference.yml specifically for moe?

**Review by mailvijayasingh — COMMENTED** [review]
Overall, looks good, left some comments. 

Lets test, dense models, training and moe 8x22B on v6e-8 before pushing

File: MaxText/layers/attentions.py [hunk 6]
```diff
@@ -68,7 +69,10 @@ class AttentionType(enum.Enum):
 BATCH = common_types.BATCH
 PREFILL_KV_BATCH = common_types.PREFILL_KV_BATCH
 KV_BATCH = common_types.KV_BATCH
+DECODE_BATCH = common_types.DECODE_BATCH
+DECODE_LENGTH = common_types.DECODE_LENGTH
 LENGTH = common_types.LENGTH
+KV_LENGTH = common_types.KV_LENGTH
```
Comment (mailvijayasingh) [inline MaxText/layers/attentions.py:75]: We will have to make changes at places where kv_quant=True or else its broken, one work around is to, push this code with an assertion that when context_parallel !=1, kv_quant should be False.
and quickly follow up with quantization changes. but if its not too hard, I would prefer them going together

File: MaxText/configs/base.yml (hunk 5 above)
Comment (RissyRan) [reply MaxText/configs/base.yml:269]: What's this config? i.e. is this context parallelism related?

File: MaxText/layers/attentions.py [hunk 7]
```diff
@@ -568,16 +582,32 @@ def apply_attention_dot(
       key = key.astype(jnp.float32)
 
     q_seq_len = query.shape[1]
+    # special sharding for decode
+    if self.config.ici_context_parallelism > 0 and q_seq_len == 1:
```
Comment (RissyRan) [inline MaxText/layers/attentions.py:586]: Nit: could we wrap this `self.config.ici_context_parallelism > 0 and q_seq_len == 1` to a helper function and reuse, and name it something like `is_context_parallelism_in_decoding()`?

File: MaxText/layers/linears.py [hunk 8]
```diff
@@ -521,7 +522,14 @@ def reshape_and_update_weights(self, weights, indices):
   def generate_masks(self, top_k_indices, softmax_probs):
     # calculate expert_capacity = (tokens_per_batch / num_experts) * capacity_factor
     batch_size, seq_len, _ = top_k_indices.shape
-    tokens_per_batch = seq_len * self.num_experts_per_tok
+    cp = self.config.ici_context_parallelism
+    if seq_len % cp != 0:
+      cp = 1
+    sub_seq = seq_len // cp
+
+    top_k_indices = jnp.reshape(top_k_indices, (batch_size, cp, sub_seq, top_k_indices.shape[2]))
```
Comment (RissyRan) [inline MaxText/layers/linears.py:530]: Could we rename `cp` as `context_partitions` or similar? And add a comment about top_k_indices shape, i.e. [batch_size, context_partition, sub_sequence, num_experts_per_tok]?

File: MaxText/layers/linears.py (hunk 4 above)
Comment (RissyRan) [reply MaxText/layers/linears.py:660]: +1, let's extract those into a helper function, and reuse. Or `get_context_partition_and_sub_seq` to return both.

File: MaxText/layers/linears.py [hunk 9]
```diff
@@ -686,25 +730,28 @@ def dense_matmul(self, inputs, gate_logits, w0_kernel, w1_kernel, wo_kernel):
...
-        )
-        if self.config.activations_in_float32:
-          intermediate_layer = intermediate_layer.astype(jnp.float32)
+              "EBNCH,EHM -> EBNCM", layer_multiply, wo_kernel, precision=matmul_precision
+          )
+
+        # intermediate_layer = nn.with_logical_constraint(
+        #     intermediate_layer,
+        #     ("activation_exp", "activation_batch_no_exp", None, "activation_embed"),
+        # )
```
Comment (RissyRan) [inline MaxText/layers/linears.py:739]: I see this block is removed

```
if self.config.activations_in_float32:
          intermediate_layer = intermediate_layer.astype(jnp.float32)
```

**Review by RissyRan — COMMENTED** [review]
Thanks for the change and improvement! Overall, LGTM! One thing I'd like to see how is the performance impact for training. When it's ready, could you help run a benchmark on 8X7b (or other model size) with FSDP + EP sharding in dropping (with and without this change)? Capturing profiles will be great! Thank you!

**Review by mailvijayasingh — APPROVED** [review]
LGTM!

File: MaxText/configs/base.yml (hunk 5 above)
Comment (suexu1025) [reply MaxText/configs/base.yml:269]: yes, it is context for context parallelism. Keep context parallelism = 1 for training will have no effect.

File: MaxText/layers/linears.py (hunk 8 above)
Comment (suexu1025) [reply MaxText/layers/linears.py:535]: please see the updated comments here.

File: MaxText/configs/base.yml (hunk 5 above)
Comment (RissyRan) [reply MaxText/configs/base.yml:269]: Oh, I was referring to `moe_config_inference.yml` config, Vijaya mentioned earlier.

File: MaxText/layers/linears.py (hunk 8 above)
Comment (RissyRan) [reply MaxText/layers/linears.py:535]: Thanks! Could we add one more line as 

`top_k_indices shape: [batch_size, context_partition, sub_sequence, num_experts_per_tok]` ?

File: MaxText/layers/linears.py [hunk 10]
```diff
@@ -686,25 +734,29 @@ def dense_matmul(self, inputs, gate_logits, w0_kernel, w1_kernel, wo_kernel):
...
-            "EBCH,EHM -> EBCM", layer_multiply, wo_kernel, precision=matmul_precision
-        )
-        intermediate_layer = nn.with_logical_constraint(
-            intermediate_layer,
-            ("activation_exp", "activation_batch_no_exp", None, "activation_embed"),
+            "EBNCH,EHM -> EBNCM", layer_multiply, wo_kernel, precision=matmul_precision
         )
         if self.config.activations_in_float32:
           intermediate_layer = intermediate_layer.astype(jnp.float32)
+        # intermediate_layer = nn.with_logical_constraint(
```
Comment (RissyRan) [inline MaxText/layers/linears.py:741]: Is this sharding config affect inference?

File: MaxText/layers/linears.py [hunk 11]
```diff
@@ -623,7 +645,9 @@ def maybe_all_gather_kernel_weight_in_expert_parallelism(self, kernel, kernel_ax
 
   def dense_matmul(self, inputs, gate_logits, w0_kernel, w1_kernel, wo_kernel):
     # gate_logits: batch, length, expert
+    # follow router_logits = shd.shard(router_logits, (None, None, None))
```
Comment (RissyRan) [inline MaxText/layers/linears.py:648]: Nit: what's the `shd.shard`? or you mean `nn.with_logical_constraint`?

**Review by RissyRan — COMMENTED** [review]
Thanks Qinwen! Would be great if you could help run a few steps of training tests with profiles. Previously we met issues in optimization of inference, and some changes have degragation of training performance.

File: MaxText/layers/linears.py (hunk 10 above)
Comment (suexu1025) [reply MaxText/layers/linears.py]: yes, I added condition that only affect inference.

File: MaxText/layers/linears.py (hunk 11 above)
Comment (suexu1025) [reply MaxText/layers/linears.py]: that's old comments from Zhiyu, cleaned it up now

File: MaxText/configs/base.yml (hunk 5 above)
Comment (suexu1025) [reply MaxText/configs/base.yml:269]: context shardings changes is general not for MOE specific. Perf is tested for MOE inference when set context. Otherwise the code will branch as current code.

File: MaxText/configs/inference.yml [hunk 12]
```diff
@@ -0,0 +1,626 @@
+# Copyright 2023 Google LLC
```
Comment (RissyRan) [inline MaxText/configs/inference.yml:1]: Nit: 2025

File: MaxText/configs/inference.yml [hunk 13]
```diff
@@ -0,0 +1,626 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# This sentinel is a reminder to choose a real run name.
+# If there is already a checkpoint under this run, that checkpoint will auto-resume.
+run_name: ""
```
Comment (RissyRan) [inline MaxText/configs/inference.yml:17]: Sorry, I made this comment in the group chat, but forget to comment here.

Instead of duplicating the whole yml, is that possible to just override the delta like this yml file? https://github.com/AI-Hypercomputer/maxtext/blob/main/MaxText/configs/dpo.yml So, we only need to maintain the delta in this file. I think it's inference team's call, but for code quality, I would prefer just have delta in this file.

cc @vipannalla @mailvijayasingh

File: MaxText/layers/linears.py [hunk 14]
```diff
@@ -556,6 +640,10 @@ def generate_masks(self, top_k_indices, softmax_probs):
     combined_expert_mask = jnp.sum(trunc_expert_mask, axis=2)
 
     # reshape & update weights
+    # softmax_probs = jnp.reshape(
```
Comment (RissyRan) [inline MaxText/layers/linears.py:643]: Shall we remove this block if not used?

File: MaxText/layers/linears.py [hunk 15]
```diff
@@ -623,7 +712,9 @@ def maybe_all_gather_kernel_weight_in_expert_parallelism(self, kernel, kernel_ax
 
   def dense_matmul(self, inputs, gate_logits, w0_kernel, w1_kernel, wo_kernel):
     # gate_logits: batch, length, expert
+    # follow router_logits non-sharded kernel
     gate_logits = nn.with_logical_constraint(gate_logits, ("activation_batch", "activation_length", None))
+    softmax_probs = jax.nn.softmax(gate_logits.astype(jnp.float32), axis=-1).astype(self.dtype)
```
Comment (RissyRan) [inline MaxText/layers/linears.py:717]: we have a this block after, is this softmax needed here?

```
    if self.config.decoder_block == "deepseek":
      top_k_weights = self.deepseek_scale_weights(top_k_weights)
    else:
      top_k_weights = jax.nn.softmax(top_k_weights.astype(jnp.float32), axis=-1).astype(self.dtype)
```

File: MaxText/layers/linears.py [hunk 16]
```diff
@@ -686,25 +811,31 @@ def dense_matmul(self, inputs, gate_logits, w0_kernel, w1_kernel, wo_kernel):
...
             combine_mask,
             precision=matmul_precision,
-        ).astype(self.dtype)
+        )
+        if output.ndim == 4:
+          output = jnp.reshape(output, (output.shape[0], output.shape[1] * output.shape[2], output.shape[3]))
       return output, loss
     else:
+      top_k_weights /= top_k_weights.sum(-1, keepdims=True)
+      weights = self.reshape_and_update_weights(top_k_weights, top_k_indices)
```
Comment (RissyRan) [inline MaxText/layers/linears.py:837]: Are those two lines needed here?

File: MaxText/layers/linears.py (hunk 15 above)
Comment (RissyRan) [reply MaxText/layers/linears.py:712]: Is this resolved?

**Review by suexu1025 — COMMENTED** [review]
address comments

File: MaxText/layers/linears.py [hunk 17]
```diff
@@ -635,49 +720,85 @@ def dense_matmul(self, inputs, gate_logits, w0_kernel, w1_kernel, wo_kernel):
...
+        dispatch_axis = ("activation_exp", "activation_batch_no_exp", None, "activation_embed")
+        mlp_axis = ("activation_exp", "activation_batch_no_exp", None, "activation_mlp")
+        dispatch_eimsum = f"BSM,BSEC -> EBCM"
+        mlp_einsum = f"EBCM,EMH -> EBCH"
+        output_einsum = f"EBCM,BSEC -> BSM"
       else:
-        loss = None
-      inputs = nn.with_logical_constraint(inputs, ("activation_batch", "activation_length", "activation_embed"))
+        # todo: try replace softmax_probs with padded weights and verify with decode acc tests 
+        softmax_probs = jax.nn.softmax(gate_logits.astype(jnp.float32), axis=-1).astype(self.dtype)
```
Comment (suexu1025) [inline MaxText/layers/linears.py:745]: currently it is branched for softmax weight for training/inference. Ideally we don't like branching too much and later on the tests couldn't catch those branch changes. I added todo comments. later on next week to test out the accuracy using padded weight, we can remove that part of  branched code.

**Review by RissyRan — APPROVED** [review]

**Review by mailvijayasingh — APPROVED** [review]
